vault/resource_plugin.go

```diff
@@ -0,0 +1,226 @@
+			consts.FieldType: {
+				Type:        schema.TypeString,
+				Description: `Type of plugin; one of "auth", "secret", or "database".`,
+				Required:    true,
+				ForceNew:    true,
+			},
+			consts.FieldName: {
+				Type:        schema.TypeString,
+				Description: "Name of the plugin.",
+				Required:    true,
+				ForceNew:    true,
+			},
+			consts.FieldVersion: {
+				Type:        schema.TypeString,
+				Description: "Semantic version of the plugin.",
+				Optional:    true,
+				ForceNew:    true,
+			},
+			fieldSHA256: {
+				Type:        schema.TypeString,
+				Description: "SHA256 sum of the plugin binary.",
+				Required:    true,
+			},
+			fieldCommand: {
+				Type:        schema.TypeString,
+				Description: "Command to execute the plugin, relative to the plugin_directory.",
+				Required:    true,
+			},
+			fieldArgs: {
+				Type:        schema.TypeList,
+				Description: "List of additional arguments to pass to the plugin.",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			fieldEnv: {
+				Type:        schema.TypeList,
+				Description: "List of additional environment variables to run the plugin with in KEY=VALUE form.",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Sensitive: true,
+			},
+			fieldOCIImage: {
+				Type:        schema.TypeString,
+				Description: "OCI image to run. If specified, setting command, args, and env will update the container's entrypoint, args, and environment variables (append-only) respectively.",
+				Optional:    true,
+			},
+			fieldRuntime: {
+				Type:        schema.TypeString,
+				Description: "Vault plugin runtime to use if oci_image is specified.",
+				Optional:    true,
+			},
+		},
+	}
+}
+
+func pluginWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := provider.GetClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	pluginType, err := api.ParsePluginType(d.Get(consts.FieldType).(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	name := d.Get(consts.FieldName).(string)
+	version := d.Get(consts.FieldVersion).(string)
+	id := fmt.Sprintf("%s/%s", pluginType, name)
+	if version != "" {
+		id = fmt.Sprintf("%s/%s", id, version)
+	}
+
+	log.Printf("[DEBUG] Writing plugin %q", id)
+	err = client.Sys().RegisterPluginWithContext(ctx, &api.RegisterPluginInput{
+		Type:     pluginType,
+		Name:     name,
+		Version:  version,
+		SHA256:   d.Get(fieldSHA256).(string),
+		Command:  d.Get(fieldCommand).(string),
+		Args:     util.ToStringArray(d.Get(fieldArgs).([]interface{})),
+		Env:      util.ToStringArray(d.Get(fieldEnv).([]interface{})),
+		OCIImage: d.Get(fieldOCIImage).(string),
+		Runtime:  d.Get(fieldRuntime).(string),
+	})
+	if err != nil {
+		return diag.Errorf("error updating plugin %q: %s", id, err)
+	}
+	log.Printf("[DEBUG] Wrote plugin %q", id)
+
+	d.SetId(id)
+
+	return nil
```

Comment:
Should we read after write here to follow Provider workflows?
```suggestion
	return pluginRead(ctx, d, meta)
```

Reply:
Oops, thanks for catching this.